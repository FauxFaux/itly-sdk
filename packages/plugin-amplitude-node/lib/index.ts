/* eslint-disable no-unused-vars, class-methods-use-this */
import {
  Event,
  Properties,
  Plugin,
  PluginLoadOptions,
} from '@itly/sdk';
import Amplitude, { AmplitudeOptions } from 'amplitude';

export { AmplitudeOptions };

/**
 * Amplitude Node Plugin for Iteratively SDK
 */
export class AmplitudePlugin extends Plugin {
  private amplitude?: Amplitude;

  constructor(
    private apiKey: string,
    private options?: AmplitudeOptions,
  ) {
    super('amplitude');
  }

  // TODO: Allow passing in an instance rather than adding
  // TODO: destination specific configuration to Options
  // constructor(instanceOrApiKey: Amplitude | string, options?: AmplitudeOptions) {
  //     if(typeof instanceOrApiKey === 'string') {
  //         this.amplitude = new Amplitude(instanceOrApiKey, options);
  //     }
  //     else {
  //         this.amplitude = instanceOrApiKey;
  //     }
  // }

  load(options: PluginLoadOptions) {
    super.load(options);
    this.amplitude = new Amplitude(this.apiKey, this.options);
  }

  async identify(userId: string, properties?: Properties) {
    const payload = {
      user_id: userId,
      user_properties: properties,
    };
    const responseLogger = this.logger!.logRequest('identify', JSON.stringify(payload));
    try {
      const response = await this.amplitude!.identify(payload);
      responseLogger.success(response);
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }

  async track(userId: string, event: Event) {
    const payload = {
      event_type: event.name,
      user_id: userId,
      event_properties: event.properties,
    };
    const responseLogger = this.logger!.logRequest('track', JSON.stringify(payload));
    try {
      const response = await this.amplitude!.track(payload);
      responseLogger.success(JSON.stringify(response));
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }
}

export default AmplitudePlugin;
