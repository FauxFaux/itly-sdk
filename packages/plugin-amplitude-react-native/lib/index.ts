/* eslint-disable no-unused-vars, class-methods-use-this */
import {
  Event,
  Properties,
  RequestLoggerPlugin,
  PluginLoadOptions,
  PluginCallOptions,
} from '@amplitude/sdk';
import { Amplitude } from '@amplitude/react-native';

export type AmplitudeOptions = { };
export type AmplitudeResponse = boolean;
export type AmplitudeCallback = (response: AmplitudeResponse) => void;

export interface AmplitudeCallOptions extends PluginCallOptions {}
export interface AmplitudeAliasOptions extends AmplitudeCallOptions {}
export interface AmplitudeIdentifyOptions extends AmplitudeCallOptions {
  callback?: AmplitudeCallback;
}
export interface AmplitudeGroupOptions extends AmplitudeCallOptions {}
export interface AmplitudePageOptions extends AmplitudeCallOptions {}
export interface AmplitudeTrackOptions extends AmplitudeCallOptions {
  callback?: AmplitudeCallback;
}

/**
 * Amplitude React Native Plugin for Iteratively SDK
 */
export class AmplitudePlugin extends RequestLoggerPlugin {
  private amplitude: Amplitude | undefined;

  private initializePromise: Promise<boolean> | undefined;

  constructor(
    private apiKey: string,
    private options?: AmplitudeOptions,
  ) {
    super('amplitude');
  }

  async load(options: PluginLoadOptions) {
    super.load(options);
    this.amplitude = this.createAmplitude();
    this.initializePromise = this.amplitude.init(this.apiKey);
    await this.initializePromise;
  }

  createAmplitude(): Amplitude {
    return Amplitude.getInstance();
  }

  async identify(userId: string | undefined, properties?: Properties, options?: AmplitudeIdentifyOptions) {
    await this.initializePromise!;

    if (userId) {
      await this.amplitude!.setUserId(userId);
    }

    if (properties) {
      const { callback } = options ?? {};
      const responseLogger = this.logger!.logRequest('identify', `${userId} ${JSON.stringify(properties)}`);
      try {
        const response = await this.amplitude!.setUserProperties(properties);
        responseLogger.success(JSON.stringify(response));
        callback?.(response);
      } catch (e) {
        responseLogger.error(e.toString());
      }
    }
  }

  async track(userId: string, { name, properties }: Event, options?: AmplitudeTrackOptions) {
    await this.initializePromise!;

    const { callback } = options ?? {};
    const responseLogger = this.logger!.logRequest('track', JSON.stringify({ name, properties }));
    try {
      const response = await this.amplitude!.logEvent(name, properties);
      responseLogger.success(JSON.stringify(response));
      callback?.(response);
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }

  async reset() {
    await this.initializePromise!;

    await this.amplitude!.setUserId(null);
    await this.amplitude!.regenerateDeviceId();
  }

  async flush() {
    await this.initializePromise!;

    await this.amplitude!.uploadEvents();
  }
}

export default AmplitudePlugin;
