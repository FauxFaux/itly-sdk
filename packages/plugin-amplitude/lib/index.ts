
/* eslint-disable no-unused-vars, class-methods-use-this */
/* eslint-disable no-restricted-syntax, no-prototype-builtins, no-continue */
import {
  ItlyEvent, ItlyProperties, ItlyPluginBase,
} from '@itly/sdk-core';

export type AmplitudeOptions = {};

export default class AmplitudeBrowserPlugin extends ItlyPluginBase {
  static ID: string = 'amplitude';

  get amplitude(): any {
    // eslint-disable-next-line no-restricted-globals
    const s: any = typeof self === 'object' && self.self === self && self;
    return s && s.amplitude;
  }

  constructor(
    private apiKey: string,
    private options?: AmplitudeOptions,
  ) {
    super();
  }

  id = () => AmplitudeBrowserPlugin.ID;

  load() {
    if (!this.amplitude) {
      // Amplitude (https://help.amplitude.com/hc/en-us/articles/115001361248-JavaScript-SDK-Installation)
      // @ts-ignore
      // eslint-disable-next-line
      (function(e,t){var n=e.amplitude||{_q:[],_iq:{}};var r=t.createElement("script") ;r.type="text/javascript";r.async=true ;r.src="https://cdn.amplitude.com/libs/amplitude-4.5.2-min.gz.js" ;r.onload=function(){if(e.amplitude.runQueuedFunctions){ e.amplitude.runQueuedFunctions()}else{ console.log("[Amplitude] Error: could not load SDK")}} ;var i=t.getElementsByTagName("script")[0];i.parentNode.insertBefore(r,i) ;function s(e,t){e.prototype[t]=function(){ this._q.push([t].concat(Array.prototype.slice.call(arguments,0)));return this}} var o=function(){this._q=[];return this} ;var a=["add","append","clearAll","prepend","set","setOnce","unset"] ;for(var u=0;u<a.length;u++){s(o,a[u])}n.Identify=o;var c=function(){this._q=[] ;return this} ;var l=["setProductId","setQuantity","setPrice","setRevenueType","setEventProperties"] ;for(var p=0;p<l.length;p++){s(c,l[p])}n.Revenue=c ;var d=["init","logEvent","logRevenue","setUserId","setUserProperties","setOptOut", "setVersionName","setDomain","setDeviceId","setGlobalUserProperties","identify", "clearUserProperties","setGroup","logRevenueV2","regenerateDeviceId","logEventWithTimestamp", "logEventWithGroups","setSessionId","resetSessionId"] ;function v(e){function t(t){e[t]=function(){ e._q.push([t].concat(Array.prototype.slice.call(arguments,0)))}} for(var n=0;n<d.length;n++){t(d[n])}}v(n);n.getInstance=function(e){ e=(!e||e.length===0?"$default_instance":e).toLowerCase() ;if(!n._iq.hasOwnProperty(e)){n._iq[e]={_q:[]};v(n._iq[e])}return n._iq[e]} ;e.amplitude=n})(window,document);
    }
    this.amplitude.getInstance().init(this.apiKey, undefined, this.options);
  }

  identify(userId: string | undefined, properties?: ItlyProperties) {
    if (userId) {
      this.amplitude.getInstance().setUserId(userId);
    }

    if (properties) {
      const identifyObject = new this.amplitude.Identify();
      for (const p in properties) {
        if (!properties.hasOwnProperty(p)) {
          continue;
        }

        identifyObject.set(p, (properties as any)[p]);
      }

      this.amplitude.getInstance().identify(identifyObject);
    }
  }

  track(userId: string | undefined, event: ItlyEvent) {
    this.amplitude.getInstance().logEvent(
      event.name,
      event.properties,
    );
  }

  reset() {
    this.amplitude.getInstance().setUserId(null);
    this.amplitude.getInstance().regenerateDeviceId();
  }
}
