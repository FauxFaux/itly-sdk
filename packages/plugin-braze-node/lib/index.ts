/* eslint-disable no-unused-vars */
import fetch from 'node-fetch';
import {
  Event, Properties, RequestLoggerPlugin, PluginLoadOptions, ResponseLogger,
} from '@itly/sdk';

export type BrazeOptions = {
  baseUrl: string;
}

/**
 * Braze Node Plugin for Iteratively SDK
 */
export class BrazePlugin extends RequestLoggerPlugin {
  private readonly userTrackUrl: string;

  constructor(
    private readonly apiKey: string,
    private readonly options: BrazeOptions,
  ) {
    super('braze');
    const baseUrl = options.baseUrl.replace(/\/$/, '');
    this.userTrackUrl = `${baseUrl}/users/track`;
  }

  load(options: PluginLoadOptions) {
    super.load(options);
  }

  async identify(userId: string, properties?: Properties) {
    const responseLogger = this.logger.logRequest('identify', `${userId}, ${JSON.stringify(properties)}`);
    const eventProperties = BrazePlugin.toBrazeProperties(properties);
    await this.postTrackerRequest(responseLogger, {
      attributes: [
        {
          ...eventProperties,
          external_id: userId,
        },
      ],
    });
  }

  async track(userId: string, { name, properties }: Event) {
    const responseLogger = this.logger.logRequest('identify', `${userId}, ${name}, ${JSON.stringify(properties)}`);
    const eventProperties = BrazePlugin.toBrazeProperties(properties);
    await this.postTrackerRequest(responseLogger, {
      events: [
        {
          ...eventProperties,
          name,
          external_id: userId,
          time: new Date().toISOString(),
          _update_existing_only: true,
        },
      ],
    });
  }

  private async postTrackerRequest(responseLogger: ResponseLogger, body: any) {
    try {
      const response = await fetch(this.userTrackUrl, {
        method: 'post',
        headers: {
          authorization: `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      });
      if (response.status < 300) {
        responseLogger.success('success');
      } else {
        responseLogger.success(`unexpected status code: ${response.status}`);
      }
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }

  private static toBrazeProperties(properties: Properties | undefined): Record<string, any> | undefined {
    return properties != null
      ? Object.fromEntries(Object.entries(properties).map(([key, value]) => [key, BrazePlugin.valueForAPI(value)]))
      : properties;
  }

  private static valueForAPI(value: any): any {
    // https://www.braze.com/docs/api/objects_filters/user_attributes_object/
    // https://www.braze.com/docs/api/objects_filters/event_object/
    // API supports doesn't support objects and (non-string) arrays.
    return (value != null && typeof value === 'object') ? JSON.stringify(value) : value;
  }
}
