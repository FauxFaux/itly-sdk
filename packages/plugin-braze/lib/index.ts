/* eslint-disable no-unused-vars, class-methods-use-this */
/* eslint-disable no-restricted-syntax */
import appboy from '@braze/web-sdk-core';
import {
  Event, Properties, RequestLoggerPlugin, PluginLoadOptions,
} from '@itly/sdk';

export type BrazeOptions = appboy.InitializationOptions;

/**
 * Braze Browser Plugin for Iteratively SDK
 */
export class BrazePlugin extends RequestLoggerPlugin {
  constructor(
    private readonly apiKey: string,
    private readonly options: BrazeOptions,
  ) {
    super('braze');
  }

  load(options: PluginLoadOptions) {
    super.load(options);
    appboy.initialize(this.apiKey, this.options);
    appboy.openSession();
  }

  identify(userId: string | undefined, properties?: Properties) {
    if (userId) {
      appboy.changeUser(userId);
    }
    if (properties != null) {
      const user = appboy.getUser();
      for (const [key, value] of Object.entries(properties)) {
        user.setCustomUserAttribute(key, BrazePlugin.valueForAPI(value));
      }
    }
  }

  track(userId: string | undefined, { name, properties }: Event) {
    if (userId) {
      appboy.changeUser(userId);
    }
    if (properties == null) {
      appboy.logCustomEvent(name);
    } else {
      const eventProperties = Object.fromEntries(
        Object.entries(properties).map(([key, value]) => [key, BrazePlugin.valueForAPI(value)]),
      );
      appboy.logCustomEvent(name, eventProperties);
    }
  }

  private static valueForAPI(value: any): any {
    // https://www.braze.com/docs/api/objects_filters/user_attributes_object/
    // https://www.braze.com/docs/api/objects_filters/event_object/
    // API supports doesn't support objects and (non-string) arrays.
    return (value != null && typeof value === 'object') ? JSON.stringify(value) : value;
  }
}
