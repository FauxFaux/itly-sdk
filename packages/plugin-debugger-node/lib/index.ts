/* eslint-disable no-unused-vars, class-methods-use-this, no-constant-condition, no-await-in-loop */
import fetch from 'node-fetch';
import {
  ItlyEvent, ItlyProperties, ItlyPluginBase, ValidationResponse,
} from '@itly/sdk-node';

export type DebuggerOptions = {
  url: string,
  debugLevel?: 'full' | 'metadataOnly',
  batchSize?: number,
  flushAt?: number,
  flushInterval?: number
};

enum TrackType {
  track = 'track',
  group = 'group',
  identify = 'identify',
  page = 'page',
}

type TrackModel = {
  type: TrackType;
  dateSent: Date;
  eventId?: string;
  eventSchemaVersion?: string;
  eventName?: string;
  properties: ItlyProperties,
  valid: boolean;
  validation: {
    details: string;
  };
};

export const wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export default class DebuggerNodePlugin extends ItlyPluginBase {
  static ID: string = 'debugger-node';

  private buffer: TrackModel[] = [];

  private config: Required<DebuggerOptions> = {
    url: '',
    debugLevel: 'full',
    batchSize: 100,
    flushAt: 10,
    flushInterval: 1000,
  };

  constructor(private apiKey: string, debuggerOptions: DebuggerOptions) {
    super();
    this.config = { ...this.config, ...debuggerOptions };

    this.startCheck();
  }

  // overrides ItlyPluginBase.id
  id = () => DebuggerNodePlugin.ID;

  // overrides ItlyPluginBase.track
  track(userId: string | undefined, event: ItlyEvent): void {
    this.push(
      this.toTrackModel(TrackType.track, event, { ...event.properties, userId }),
    );
  }

  // overrides ItlyPluginBase.validationError
  validationError(validationResponse: ValidationResponse, event: ItlyEvent): void {
    this.push(
      this.toTrackModel(TrackType.track, event, event.properties, validationResponse),
    );
  }

  // overrides ItlyPluginBase.group
  group(userId: string | undefined, groupId: string, properties?: ItlyProperties): void {
    this.push(
      this.toTrackModel(TrackType.group, undefined, { ...properties, userId, groupId }),
    );
  }

  // overrides ItlyPluginBase.identify
  identify(userId?: string, properties?: ItlyProperties): void {
    this.push(
      this.toTrackModel(TrackType.identify, undefined, { ...properties, userId }),
    );
  }

  // overrides ItlyPluginBase.page
  page(userId?: string, category?: string, name?: string, properties?: ItlyProperties): void {
    this.push(
      this.toTrackModel(TrackType.page, undefined, {
        ...properties, userId, category, name,
      }),
    );
  }

  private toTrackModel(type: TrackType, event?: ItlyEvent, properties?: ItlyProperties,
    validation?: ValidationResponse): TrackModel {
    const model = {
      type,
      dateSent: new Date(new Date().getTime()),
      eventId: event ? event.id : undefined,
      eventSchemaVersion: event ? event.version : undefined,
      eventName: event ? event.name : undefined,
      properties: {},
      valid: true,
      validation: {
        details: '',
      },
    };

    if (properties) {
      if (this.config.debugLevel === 'full') {
        model.properties = properties;
      } else {
        model.properties = Object.keys(properties).reduce((o, key) => ({ ...o, [key]: null }), {});
      }
    }

    if (validation) {
      model.valid = validation.valid;
      if (this.config.debugLevel === 'full') {
        model.validation.details = `${validation.pluginId}: ${validation.message}`;
      }
    }

    return model;
  }

  private async flush() {
    while (this.buffer.length > 0) {
      const objects = this.buffer.splice(0, this.config.batchSize);
      await fetch(this.config.url, {
        method: 'post',
        headers: {
          authorization: `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          objects,
        }),
      });
    }
  }

  private push(model: TrackModel) {
    this.buffer.push(model);
    if (this.buffer.length >= this.config.flushAt) {
      this.flush();
    }
  }

  private async startCheck() {
    while (true) {
      await this.check();
    }
  }

  private async check() {
    if (this.buffer.length > 0) {
      this.flush();
    }
    await wait(this.config.flushInterval);
  }
}
