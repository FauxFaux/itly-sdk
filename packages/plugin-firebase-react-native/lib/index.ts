/* eslint-disable no-unused-vars, class-methods-use-this */
import analytics from '@react-native-firebase/analytics';
import {
  Event, Properties, RequestLoggerPlugin,
} from '@amplitude/sdk';

export type FirebaseOptions = {}

/**
 * Firebase React Native Plugin for Iteratively SDK
 */
export class FirebasePlugin extends RequestLoggerPlugin {
  constructor(
    private readonly options?: FirebaseOptions,
  ) {
    super('firebase');
  }

  async identify(userId: string | undefined, properties?: Properties) {
    const responseLogger = this.logger!.logRequest('identify', `${userId}, ${JSON.stringify(properties)}`);
    try {
      if (userId) {
        await analytics().setUserId(userId);
      }
      if (properties != null) {
        const identifyProperties = Object.fromEntries(
          Object.entries(properties).map(([key, value]) =>
            [key, (value == null || typeof value === 'string') ? value : JSON.stringify(value)]),
        );
        await analytics().setUserProperties(identifyProperties);
      }
      responseLogger.success('done');
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }

  async page(userId?: string, category?: string, name?: string, properties?: Properties) {
    const responseLogger = this.logger!.logRequest('page', `${userId}, ${category}, ${name}, ${JSON.stringify(properties)}`);
    try {
      if (userId) {
        await analytics().setUserId(userId);
      }
      await analytics().logScreenView({
        screen_class: category,
        screen_name: name,
      });
      responseLogger.success('done');
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }

  async track(userId: string | undefined, { name, properties }: Event) {
    const responseLogger = this.logger!.logRequest('track', `${userId}, ${name}, ${JSON.stringify(properties)}`);
    try {
      if (userId) {
        await analytics().setUserId(userId);
      }
      await analytics().logEvent(
        name,
        properties,
      );
      responseLogger.success('done');
    } catch (e) {
      responseLogger.error(e.toString());
    }
  }

  async reset() {
    await analytics().setUserId(null);
  }
}

export default FirebasePlugin;
