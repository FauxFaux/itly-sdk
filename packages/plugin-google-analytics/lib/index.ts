/* eslint-disable no-restricted-globals */
/* eslint-disable no-underscore-dangle */
/* eslint-disable no-unused-vars, class-methods-use-this */
import {
  Event, Properties, RequestLoggerPlugin, PluginLoadOptions, PluginCallOptions,
} from '@itly/sdk';

type GoogleAnalyticsOptions = {
  gtag?: Gtag.Gtag;
  measurementId: string;
}

export interface GoogleAnalyticsCallOptions extends PluginCallOptions { }
export interface GoogleAnalyticsPageOptions extends PluginCallOptions { }

export class GoogleAnalyticsPlugin extends RequestLoggerPlugin {
  private _gtag?: Gtag.Gtag;

  get gtag(): Gtag.Gtag {
    this._gtag = this._gtag || self.gtag;
    if (!this._gtag) throw new Error('No gtag provided to constructor, or available on window');
    return this._gtag;
  }

  constructor(private opts: GoogleAnalyticsOptions) {
    super('google-analytics');
  }

  load(options: PluginLoadOptions) {
    super.load(options);
    if (this.opts.gtag) {
      this._gtag = this.opts.gtag;
    } else if ('gtag' in self) {
      this._gtag = self.gtag;
    } else {
      this.initializeGtag(this.opts.measurementId);
    }
  }

  identify(userId: string | undefined, properties?: Properties | undefined) {
    this.gtag('config', this.opts.measurementId, { userId, ...properties });
  }

  track(userId: string | undefined, { name, properties }: Event, options?: GoogleAnalyticsCallOptions) {
    if (userId) { // On browser, userId will always be undefined
      this.identify(userId);
    }
    this.gtag('event', name, { ...properties, event_callback: options?.[this.id]?.callback });
    // TODO: look into if we need to stringify non-leaf values
  }

  page(
    userId: string | undefined,
    category: string | undefined,
    name: string | undefined,
    properties: Properties | undefined,
    options?: GoogleAnalyticsPageOptions,
  ) {
    if (userId) this.identify(userId);
    this.gtag('event', 'page_view', {
      page_title: name,
      page_path: self.location.href,
      event_category: category,
      send_to: this.opts.measurementId,
      event_callback: options?.[this.id]?.callback,
      ...properties,
    });
  }

  private initializeGtag(measurementId: string) {
    return new Promise((resolve) => {
      (self as any).dataLayer = (self as any).dataLayer || [];
      // eslint-disable-next-line prefer-rest-params
      self.gtag = function gtag() { (self as any).dataLayer.push(arguments); };
      self.gtag('js', new Date());
      self.gtag('config', measurementId);

      const $script = document.createElement('script');
      $script.setAttribute('async', 'async');
      $script.setAttribute('src', `https://www.googletagmanager.com/gtag/js?id=${measurementId}`);
      $script.onload = resolve;
      document.head.appendChild($script);
    });
  }
}

export default GoogleAnalyticsPlugin;
