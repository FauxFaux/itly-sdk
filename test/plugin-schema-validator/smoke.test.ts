/* eslint-disable import/no-unresolved,  no-unused-vars, global-require, import/extensions */
import { ItlyOptions, ItlyPlugin } from '../../packages/sdk-core/lib';
import SchemaValidator from '../../packages/plugin-schema-validator/lib';
import CustomPlugin from '../../test/src/CustomPlugin';

const testSchemas = require('../data/basic-schema.json');

type TestParams = {
  name: string;
  options: ItlyOptions,
};

const userId = 'test-user-id';
let spyConsoleLog: jest.SpyInstance;

const plugins: ItlyPlugin[] = [
  new SchemaValidator(testSchemas, (validation, event, schema) => {
    // eslint-disable-next-line no-console
    console.log(
      `SchemaValidator validationError() event='${event.name}' message='${validation.message}' schema=`,
      JSON.stringify(schema),
    );
  }),
  new CustomPlugin(),
];

const testParams: TestParams[] = [
  {
    name: 'load, track, validate - validationOptions=DEFAULT',
    options: {
      environment: 'production',
      plugins,
    },
  },
  {
    name: 'load, track, validate - validationOptions={trackInvalid: true}',
    options: {
      environment: 'production',
      plugins,
      validationOptions: {
        disabled: false,
        trackInvalid: true,
        errorOnInvalid: false,
      },
    },
  },
  {
    name: 'load, track, validate - validationOptions={errorOnInvalid: true}',
    options: {
      environment: 'production',
      plugins,
      validationOptions: {
        disabled: false,
        trackInvalid: false,
        errorOnInvalid: true,
      },
    },
  },
];

beforeEach(() => {
  jest.resetModules();
  spyConsoleLog = jest.spyOn(console, 'log');
});

afterEach(() => {
  spyConsoleLog.mockRestore();
});

test.each(testParams.map((test) => [test.name, test]) as any[])('%s',
  async (name: string, { options }: TestParams) => {
    const { default: itly } = require('../../packages/sdk-core/lib');

    itly.load(options);

    itly.track(userId, {
      name: 'Event No Properties',
      id: '26af925a-be3a-40e5-947d-33da66a5352f',
      version: '1-0-0',
      properties: {},
    });

    itly.track(userId, {
      name: 'Event With All Properties',
      id: '311ba144-8532-4474-a9bd-8b430625e29a',
      version: '1-0-0',
      properties: {
        requiredString: 'A required string',
        requiredNumber: 2.0,
        requiredInteger: 42,
        requiredEnum: 'Enum1',
        requiredBoolean: false,
        requiredConst: 'some-const-value',
        requiredArray: ['required', 'array'],
        optionalString: 'I\'m optional!',
      },
    });

    try {
      itly.track(userId, {
        name: 'EventMaxIntForTest',
        id: 'aa0f08ac-8928-4569-a524-c1699e7da6f4',
        version: '1-0-0',
        properties: {
          intMax10: 20,
        },
      });
    } catch (e) {
      // do nothing
      // eslint-disable-next-line no-console
      console.log('Caught validation error.', e.message);
    }

    expect(spyConsoleLog.mock.calls).toMatchSnapshot();
  });
