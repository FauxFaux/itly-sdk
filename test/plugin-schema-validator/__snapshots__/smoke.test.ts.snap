// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load, track, validate - validationOptions={errorOnInvalid: true} 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "SchemaValidator validationError() event='EventMaxIntForTest' message='Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.' schema=",
    "{\\"$id\\":\\"https://iterative.ly/company/77b37977-cb3a-42eb-bce3-09f5f7c3adb7/event/aa0f08ac-8928-4569-a524-c1699e7da6f4/version/1.0.0\\",\\"$schema\\":\\"http://json-schema.org/draft-07/schema#\\",\\"title\\":\\"EventMaxIntForTest\\",\\"description\\":\\"Event to test schema validation\\",\\"type\\":\\"object\\",\\"properties\\":{\\"intMax10\\":{\\"description\\":\\"property to test schema validation\\",\\"type\\":\\"integer\\",\\"maximum\\":10}},\\"additionalProperties\\":false,\\"required\\":[\\"intMax10\\"]}",
  ],
  Array [
    "CustomPlugin: ",
    "validationError() event='EventMaxIntForTest' plugin=schema-validator message=Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.",
  ],
  Array [
    "Caught validation error.",
    "Validation Error: Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.",
  ],
]
`;

exports[`load, track, validate - validationOptions={trackInvalid: true} 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "SchemaValidator validationError() event='EventMaxIntForTest' message='Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.' schema=",
    "{\\"$id\\":\\"https://iterative.ly/company/77b37977-cb3a-42eb-bce3-09f5f7c3adb7/event/aa0f08ac-8928-4569-a524-c1699e7da6f4/version/1.0.0\\",\\"$schema\\":\\"http://json-schema.org/draft-07/schema#\\",\\"title\\":\\"EventMaxIntForTest\\",\\"description\\":\\"Event to test schema validation\\",\\"type\\":\\"object\\",\\"properties\\":{\\"intMax10\\":{\\"description\\":\\"property to test schema validation\\",\\"type\\":\\"integer\\",\\"maximum\\":10}},\\"additionalProperties\\":false,\\"required\\":[\\"intMax10\\"]}",
  ],
  Array [
    "CustomPlugin: ",
    "validationError() event='EventMaxIntForTest' plugin=schema-validator message=Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='EventMaxIntForTest' properties={\\"intMax10\\":20}",
  ],
]
`;

exports[`load, track, validate - validationOptions=DEFAULT 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event No Properties' properties={}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='test-user-id' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "SchemaValidator validationError() event='EventMaxIntForTest' message='Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.' schema=",
    "{\\"$id\\":\\"https://iterative.ly/company/77b37977-cb3a-42eb-bce3-09f5f7c3adb7/event/aa0f08ac-8928-4569-a524-c1699e7da6f4/version/1.0.0\\",\\"$schema\\":\\"http://json-schema.org/draft-07/schema#\\",\\"title\\":\\"EventMaxIntForTest\\",\\"description\\":\\"Event to test schema validation\\",\\"type\\":\\"object\\",\\"properties\\":{\\"intMax10\\":{\\"description\\":\\"property to test schema validation\\",\\"type\\":\\"integer\\",\\"maximum\\":10}},\\"additionalProperties\\":false,\\"required\\":[\\"intMax10\\"]}",
  ],
  Array [
    "CustomPlugin: ",
    "validationError() event='EventMaxIntForTest' plugin=schema-validator message=Passed in EventMaxIntForTest properties did not validate against your tracking plan. \`properties.intMax10\` should be <= 10.",
  ],
]
`;
